website/addons/dryad/utils.py

```diff
@@ -0,0 +1,272 @@
+import re
+import requests
+import xml.dom.minidom
+from bs4 import BeautifulSoup
+from website.addons.dryad import settings as dryad_settings
+
+
+class DryadConnection:
+    """
+        DryadConnection is a proxy class for connecting to the Dryad API.
+
+        note: This will be expanded in the v2 API.
+
+        A proxy class was needed for unit testing so that the connection object
+        in each of the methods in this module can be replaced with a non-http
+        using connection.
+    """
+
+    def get(self, url='localhost', params={}):
+        if params:
+            r = requests.get(url, params=params)
+        else:
+            r = requests.get(url)
+        return r
+
+    def post(self, url):
+        r = requests.post(url)
+        return r
+
+
+class DryadRepository:
+    def __init__(self, connection=DryadConnection()):
+        self.connection = connection
+
+    def get_dryad_metadata(self, doi='doi:10.5061/dryad.1850'):
+        """ Retrieves the metadata of a dryad item in the form of xml.Document
+        :param doi: Dryad DOI in the form of "doi:10.5061/dryad.XXXX"
+        :type doi: string
+        :returns:  xml.dom.minidom.Document -- xml form of the dryad metadata
+        :raises: urllib.error.HTTPError
+        """
+        resp = self.connection.get(dryad_settings.DRYAD_DATAONE_METADATA.format(doi))
+        # The below encoding is mandatory
+        return xml.dom.minidom.parseString(resp.text.encode('utf-8'))
+
+    def get_dryad_page(self, doi='doi:10.5061/dryad.1850'):
+        """
+            Retrieves the html of the dryad page for a package. As some of the
+            object metadata is served directly to the page, this kludge is used
+            to scrape that data off the page.
+        :param doi: Dryad DOI in the form of "doi:10.5061/dryad.XXXX"
+        :type doi: string
+        :returns:  beautifulsoup4.BeautifulSoup -- soup form of the dryad metadata
+        :raises: urllib.error.HTTPError
+        """
+        resp = self.connection.get(dryad_settings.DRYAD_WEB_API.format(doi))
+        # The below encoding is mandatory
+        return BeautifulSoup(resp.text.encode('utf-8'))
```

Comment:
We tend to try to avoid `BeautifulSoup` because it's kind of… slow. Also, it means you’re parsing HTML, which means:
a) The file format could be changed any time as they redesign the front end, or,
b) You’re actually parsing an acrhival format that is xml, in which case `lxml` is preferred. It’s fast, purpose designed, and has tools for extracting data from files.